File: docs/site/samples/python/examples/guides_derive_replace_body.py
"""Derive guide — replace only the root body (the default arm) with ``derive``.

Standalone — the docs page displays this file verbatim; run it directly to check.
"""

from prompting_press import Prompt
from pydantic import BaseModel, field_validator

assistant_yaml = """
name: assistant
role: system
body: "You are a support assistant for {{ company }}. Keep your replies under {{ max_words }} words."
variables:
  company: { type: string, trusted: true }
  max_words: { type: integer, trusted: true }
"""


class AssistantVars(BaseModel):
    company: str
    max_words: int

    @field_validator("max_words")
    @classmethod
    def _at_least_one(cls, v: int) -> int:
        if v < 1:
            raise ValueError("max_words must be at least 1")
        return v


def main() -> None:
    assistant = Prompt.from_yaml(assistant_yaml)

    brief_assistant = assistant.derive(
        {"body": "You are a support assistant for {{ company }}."}
    )

    result = brief_assistant.render(
        AssistantVars, data={"company": "Acme Robotics", "max_words": 50}
    )
    assert result.text == "You are a support assistant for Acme Robotics."


if __name__ == "__main__":
    main()

File: docs/site/samples/rust/examples/guides_derive_replace_body.rs
//! Derive guide — replace only the root body (the default arm) with `derive`.
//! Standalone — `cargo run --example guides_derive_replace_body`.

use garde::Validate;
use prompting_press::{GuardConfig, Prompt, PromptOverlay};
use serde::Serialize;

#[derive(Serialize, Validate)]
struct AssistantVars {
    #[garde(length(min = 1))]
    company: String,
    #[garde(range(min = 1))]
    max_words: i64,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let assistant_yaml = r#"
name: assistant
role: system
body: "You are a support assistant for {{ company }}. Keep your replies under {{ max_words }} words."
variables:
  company: { type: string, trusted: true }
  max_words: { type: integer, trusted: true }
"#;
    let assistant = Prompt::from_yaml(assistant_yaml)?;

    let brief_assistant = assistant.derive(PromptOverlay {
        body: Some("You are a support assistant for {{ company }}.".to_string()),
        ..Default::default()
    })?;

    let vars = AssistantVars {
        company: "Acme Robotics".into(),
        max_words: 50,
    };
    let result = brief_assistant.render(&vars, None, &GuardConfig::default(), false)?;
    assert_eq!(result.text, "You are a support assistant for Acme Robotics.");
    Ok(())
}
